### PR TITLE
Change build to docker multi-stage builds

### PR DESCRIPTION
**Pull request by mkam423**
Moved the main.go out of examples to its own respective cmd folder. Updated the Dockerfiles so that we are using docker multi-stage builds and removed the respective build targets in the Makefile.

No actual logic of the code was changed.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -2,23 +2,6 @@ VERSION := $(shell git describe --dirty=-dirty --always)
```
Comment (kd7lxl) [inline Makefile]: I notice you use numbered releases for this project. Do you want to add `--tags` to the `VERSION` to include that release number?

File: Makefile [hunk 2]
```diff
@@ -91,10 +63,10 @@ test:
 
 # --- Kuberenetes deployment ---
 # Deploy the pubsub service in kubernetes
-deploy-server: 
+deploy-server:
 	cat deploy/pubsub.yaml | sed 's/\$$PUBSUB_NAMESPACE/$(PUBSUB_NAMESPACE)/g' | @kubectl create -f -
 
-# Deployes the hello example publisher and subscriber 
+# Deployes the hello example publisher and subscriber
```
Comment (kd7lxl) [inline Makefile]:
```suggestion
 # Deploys the hello example publisher and subscriber
```

File: docker/Dockerfile [hunk 3]
```diff
@@ -1,13 +1,23 @@
-FROM alpine:3.5
+# build the server binary
+FROM golang:1.10.0 AS builder
```
Comment (kd7lxl) [inline docker/Dockerfile]: Why not 1.11?

File: docker/Dockerfile [hunk 4]
```diff
@@ -1,13 +1,23 @@
...
-COPY bin/pubsub /usr/local/bin/pubsub
+
+# run the server binary
+FROM alpine:latest AS runner
+WORKDIR /usr/local/bin
+
+COPY --from=builder /go/src/github.com/infobloxopen/atlas-pubsub/pubsub .
 
 # Allows to verify certificates
 RUN apk update --no-cache && apk add --no-cache ca-certificates
```
Comment (kd7lxl) [inline docker/Dockerfile]:
```suggestion
 RUN apk add --no-cache ca-certificates
```

`apk update --no-cache` doesn't do anything, since it doesn't cache the files it just updated.

File: docker/Dockerfile (hunk 3 above)
Comment (mkam423) [reply docker/Dockerfile]: no reason. Changing now.

File: docker/Dockerfile (hunk 4 above)
Comment (mkam423) [reply docker/Dockerfile]: I'm a little confused by your comment. Didn't know too much about it so I have a resource here:
https://wiki.alpinelinux.org/wiki/Alpine_Linux_package_management#Update_the_Package_list

TLDR: It says it does cache files and that it's good to run updates before adds.

Comment (kd7lxl) [reply docker/Dockerfile]: Sorry, I meant to reference the `--no-cache` documentation, but I forgot to paste.

https://github.com/gliderlabs/docker-alpine/blob/master/docs/usage.md#disabling-cache

`--no-cache` implies `update` and deletes the repo data (what update populates) when it's done.

File: docker/Dockerfile [hunk 5]
```diff
@@ -1,13 +1,23 @@
-FROM alpine:3.5
+# build the server binary
+FROM golang:1.11.0 AS builder
```
Comment (kd7lxl) [inline docker/Dockerfile]:
```suggestion
 FROM golang:1.11 AS builder
```
The current stable release is `1.11.4`. You can do `1.11` to catch all the patches on the 1.11 version.

Comment (mkam423) [reply docker/Dockerfile]: gotchu
